File: java/1239. Maximum Length of a Concatenated String with Unique Characters.java
import java.util.Arrays;
import java.util.List;

class Solution {
    public int maxLength(List<String> arr) {
        int[] dp = new int[arr.size()];
        Arrays.fill(dp, -1);
        return helper(arr, 0, "",dp);
    }

    public int helper(List<String> arr, int index,String currStr, int[] dp){
        //if index out of range, reutrn current length
        if(index >= arr.size()){
            return currStr.length();
        }
        //if current arr element does not break uniqueness, check max(contain arr[index], currStr)
        int max_len = currStr.length();
        if(isUnique(arr.get(index), currStr)){
            String newStr = currStr + arr.get(index);
            max_len = Math.max(max_len,helper(arr,index+1,newStr,dp));
        }
        //now check previous max and notContain(arr[index])(prevMax, notContain)
        //store in dp[index], return dp[index]
        max_len = Math.max(max_len, helper(arr,index+1,currStr,dp));
        dp[index] = max_len;
        return dp[index];
        

    }
    public boolean isUnique(String str1, String str2){
        for(int i = 0; i < str1.length(); i++){
            String strCheck = String.valueOf(str1.charAt(i));
            if(str2.contains(strCheck)){
                return false;
            }
            else{
                str2 = str2 + strCheck;
            }
        }
        return true;
    }

}

File: python/1239. Maximum Length of a Concatenated String with Unique Characters.py
from ast import List


class Solution:
    def maxLength(self, arr: List[str]) -> int:
        dp = [-1 for i in range(len(arr))]
        return self.helper(arr,0,dp,set(), "")
    def helper(self, arr, index, dp, store, currStr):
        if index >= len(arr):
            return len(currStr)
        else:
            notContain = self.helper(arr,index+1,dp,store.copy(),currStr)
            contain = len(currStr)
            isUnique = True
            tempStore = store.copy()
            tempCurrStr = currStr
            for ch in arr[index]:
                if ch in tempStore:
                    isUnique = False
                    break
                else:
                    tempStore.add(ch)
                    tempCurrStr += ch
            if isUnique:
                contain = self.helper(arr,index+1,dp,tempStore, tempCurrStr)
            len1 = notContain
            len2 = contain
            len3 = len(currStr)
            if len1 >= len2 and len1 >= len3:
                dp[index] = len1
                return dp[index]
            elif len2 >= len1 and len2 >= len3:
                dp[index] = len2
                return dp[index]
            else:
                dp[index] = len3
                return dp[index]

# Time Complexity is O(2^n) where n is the length of arr. With dp, we use bottom up (tabulation), and the recursion depth is O(n), and we use recursion in two different brancehs, one with containing the current index, and without containing the current index, so the time complexity becomes O(2^n).
# Space Complexity is O(n) where n is the length of the 